internal/backup/volumesnapshot_action.go

```diff
@@ -149,13 +153,78 @@ func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, back
 
 	vsMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&vs)
 	if err != nil {
-		return nil, nil, errors.WithStack(err)
+		return nil, nil, "", nil, errors.WithStack(err)
 	}
 
 	p.Log.Infof("Returning from VolumeSnapshotBackupItemAction with %d additionalItems to backup", len(additionalItems))
 	for _, ai := range additionalItems {
 		p.Log.Debugf("%s: %s", ai.GroupResource.String(), ai.Name)
 	}
 
-	return &unstructured.Unstructured{Object: vsMap}, additionalItems, nil
+	// The operationID is of the form <namespace>/<volumesnapshot-name>
+	operationID := vs.Namespace + "/" + vs.Name
+	itemToUpdate := []velero.ResourceIdentifier{
```

Comment:
Since you're returning the current resource (the VolumeSnapshot) to update during finalize, this plugin will be called again, so at the beginning of the Execute func, you need to check backup.Phase and if it's Finalizing or FinalizingPartiallyFailed, then just return the input item without any changes or actions since all of the Execute actions already happened during InProgress.

Reply:
Thanks. Added the checking logic.